Production/src/TauJet.cc

```diff
@@ -353,6 +380,8 @@ void TauJetBuilder::Build()
             if(tauJet.fatJet)
                 pfCandDesc.subJetDaughter = GetMatchedSubJetIndex(*tauJet.fatJet, pfCand);
             tauJet.cands.push_back(pfCandDesc);
+	    if(SVidx>0)
```

Comment:
Two things here: 
* I think vertex index can be 0 -> should it be `SVidx>=0`?
* I am not sure that the current data structure of storing SVs is intuitive and efficient. Correct me if I'm wrong, my understanding is that for each `tauJet` there's a collection of matched PF candidates required to come from SV (see comment above) and for _each_ of them associated SV is stored into a separate vector. I'd say that it will create many duplications, plus SVs currently do not carry information about PF candidates associated to it (see `FillSV()` function). That means it won't be possible to build the link between PF candidates and SVs afterwards.

So I would propose the following logic: for each `tauJet` object, store a collection of SVs, where every SVs is required to have at least one associated PF candidate coming from `tauJet`. Then, for all PF candidates in `tauJet` store an additional variable which is the index of associated SV (-1 if it doesn't exist). Alternatively, one can store for each SV a vector of associated PF candidates' indices. Does it sound reasonable?

Reply:
I agree, SVidx can be 0 thank you.

So the idea was to check for each PF candidate inside or outside the cone (as we discussed earlier) if it has a matched SV. If so, (i.e ```SVidx>=0```) we keep this SV in the collection. But you are right, if a SV has several associated PFCandidates it will be stored as many times. So maybe this piece of code can stay and I will include a small function in the ```FillSV()``` function of the TauTupleProducer that both check for duplicates and keep the associated PF candidate index in a vector, as you suggested.

Reply:
> Perhaps the fillSVMatched() function is useless and can be replaced by the line

Yes indeed, because `collectSV()` is already called before, so it is enough only to append the vertex with `SVidx` to `tauJet.secondVertices` (if `SVidx >=0`) 

> So maybe this piece of code can stay and I will include a small function in the FillSV() function of the TauTupleProducer that both check for duplicates and keep the associated PF candidate index in a vector, as you suggested.

Actually, I am thinking that maybe it is simpler to store all the SVs in the event without doing any duplicate checks. Because if we want to keep all PF candidates with at least one SV outside of the cone, this necessarily implies that we would need to store quite a lot of SVs anyway (but honestly it's hard for me to tell a priori, if SVs are mostly located outside of the cone or not).

So I view it smth like this: in `TauJet.cc` PF candidates are kept if `hasMatch(pfCand.polarP4()) || SVidx>=0` (so as it is now). No information about associated SV is stored, because actually I realised that in theory one PF candidate can be associated with several SVs. Then outside of this PF loop, `tauJet.secondVertices` are filled with basically all SVs in the event. And in `TauTupleProducer.cc` in `FillSV()` it is like it is now, plus for each SV a vector of PF indices associated to it is stored (in general, without any requirement on tau cone matching etc). But here I am actually not sure, is it possible to implement it like this?

And also I have a concern that this may increase the computation time/tuple size. But I also don't have any intuitive estimate here, so maybe we could try and see how it goes..

Reply:
We need to converge with few things related to the logic of filling SV:
1) what about SVs that match to the TauJet but does not have associated pfCand within this cone, i.e: I propose, to iterate through SVs and fill them is they have PfCand or match cone.
2) Is situation when one pfCand is associated to several SVs is possible. I understand that on the level of matching combinations it is allowed, but will such "fake" SVs be saved in the collection in this case? -> if not, I propose to store index of the SVs for every PfCand and not vice versa.
3) Also, duplicate SVs in the collection would be very bad and would complicate life for the further steps.

I would do it in the following way
```c++

std::set<size_t> matched_sv;

for(size_t pfCandIndex = 0; pfCandIndex < cands_.size(); ++pfCandIndex) {
   ...
   auto SVidx = collectSV(pfCand);
   if(!hasMatch(pfCand.polarP4()) && SVidx<0) continue;
   ...
   if(SVidx >= 0) matched_sv.insert(SVidx);
   // add the index of SV to PfCand and only them push back
   tauJet.cands.push_back(pfCandDesc);
}
...
for(size_t sv_i = 0; sv_i < secondVertices_.size(); ++sv_i) {
   if( !hasMatch(secondVertices_[sv_i].polarP4()) && matched_sv.count(sv_i)) continue;
   out_SVcol.emplace_back(in_SVcol.at(index), index);
   matched_sv.erase(sv_i) 
}

//assert that matched_sv is empty

```
@kandrosov would you agree with such logic?

Reply:
Sorry, previous solution does not account for correct preserving of indexes of SVs for PfCand for later reference (I am not sure that we need it), but I guess to preserve correct indexes the only option would be putting SVs in the vector before we are putting SV and later putting the rests, probably it is a bit less efficient, but:
```c++
std::map<size_t, size_t> sv_tuple_idxs;

for(size_t pfCandIndex = 0; pfCandIndex < cands_.size(); ++pfCandIndex) {
   ...
   auto SVidx = collectSV(pfCand);
   if(!hasMatch(pfCand.polarP4()) && SVidx<0) continue;
   ...
   pfCandDesc.sv_index = -1; //some default value
   if(SVidx >= 0) {
      if(!sv_tuple_idxs.count(SVidx)) {
         sv_tuple_idxs.insert({SVidx, sv_tuple_idxs.size()};
         tauJet.secondVertices.push_back(secondVertices_.at(SVidx));
      }
      pfCandDesc.sv_index = sv_tuple_idxs.at(SVidx);
   }
   tauJet.cands.push_back(pfCandDesc);
}
...
for(size_t sv_i = 0; sv_i < secondVertices_.size(); ++sv_i) {
   if(matched_sv.count(sv_i)) continue;
   if(!hasMatch(secondVertices_.at(sv_i).polarP4())) continue;
   tauJet.secondVertices.push_back(secondVertices_.at(sv_i));
}
```